### PR TITLE
UGC safeness integration tests

### LINKED ISSUE #4830: Integration tests to verify lakeFS is safe while GC is running

### PR DESCRIPTION
**Pull request by eden-ohana**
closes #4830 

haven't added the copy object test yet - waiting for the PR

### CODE REVIEW

File: clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala [hunk 1]
```diff
@@ -146,7 +148,8 @@ object UncommittedGarbageCollector {
     try {
       if (shouldMark) {
         // Read objects directly from object storage
-        val dataDF = listObjects(storageNamespace, cutoffTime)
+        val (dataDF, startListTime) = listObjects(storageNamespace, cutoffTime)
+        listTime = startListTime
```
Comment (itaiad200) [inline clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala:152]:
```suggestion
        val (dataDF, listTime) = listObjects(storageNamespace, cutoffTime)
```

File: clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala [hunk 2]
```diff
@@ -270,6 +276,7 @@ object UncommittedGarbageCollector {
       "success" -> success,
       "first_slice" -> firstSlice,
       "start_time" -> DateTimeFormatter.ISO_INSTANT.format(startTime),
+      "start_list_time" -> DateTimeFormatter.ISO_INSTANT.format(startTime),
```
Comment (itaiad200) [inline clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala:279]: Probably a dumb question, but why do we need both? What does `start_list_time` give us that `start_time` don't?

**Review by itaiad200 — COMMENTED** [review]
I only noticed it's a draft PR half way through. Still posting 2 comments as I believe they might help.

File: esti/ugc_test.go [hunk 3]
```diff
@@ -214,16 +220,177 @@ func listUnderlyingStorage(t *testing.T, ctx context.Context, s3Client *s3.S3, b
...
+		t.Fatalf("Failed to get start list time from ugc report %s", err)
+	}
+
+	// verify that all objects are deleted, if needed
+	for _, obj := range objects {
+		if strings.Contains(obj.Key, "/_lakefs/retention/gc/uncommitted/") || strings.Contains(obj.Key, "dummy") {
+			continue
+		}
+
+		if obj.LastModified.Before(startListTime) {
```
Comment (itaiad200) [inline esti/ugc_test.go:356]: OK now I see what this test is all about. You upload many objects and delete them all. Some are written before the cutoff and some after the cutoff. Here you validate that all objects before the cutoff are deleted which is great.
I think the previous test checks that too and also that committed/uncommitted objects remain (i.e. not deleted).
Here's what I think is still missing:
1. Validation that deleted objects that were created after the cutoff are not deleted.
2. Validation that copied objects are not deleted (i.e. copy table works). It's a difficult one, you need to copy while deleting the source object, at the exact time that the uncommitted objects are being listed for UGC, yeah..

**Review by itaiad200 — CHANGES_REQUESTED** [review]
I enjoyed reading this, very well structured code. 
Requesting changes mostly for the first bullet of the missing test coverage.

File: clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala [hunk 4]
```diff
@@ -270,6 +280,7 @@ object UncommittedGarbageCollector {
       "success" -> success,
       "first_slice" -> firstSlice,
       "start_time" -> DateTimeFormatter.ISO_INSTANT.format(startTime),
+      "start_list_time" -> DateTimeFormatter.ISO_INSTANT.format(cutoffTime),
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala]: done

File: esti/ugc_test.go [hunk 5]
```diff
@@ -181,7 +187,7 @@ func validateUncommittedGC(t *testing.T) {
 	}
 }
 
-func listRepositoryUnderlyingStorage(t *testing.T, ctx context.Context, repo string) []string {
+func listRepositoryUnderlyingStorage(t *testing.T, ctx context.Context, repo string) ([]Object, block.QualifiedPrefix) {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/ugc_test.go:190]: done

File: esti/ugc_test.go [hunk 6]
```diff
@@ -214,16 +220,177 @@ func listUnderlyingStorage(t *testing.T, ctx context.Context, s3Client *s3.S3, b
...
+		Staging: api.StagingLocation{
+			PhysicalAddress: getPhysicalResp.JSON200.PhysicalAddress,
+			Token:           getPhysicalResp.JSON200.Token,
+		},
+	})
+	if err != nil {
+		t.Fatalf("Failed to link physical address %s", err)
+	}
+
+	s3Client := newDefaultS3Client()
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/ugc_test.go:326]: done

File: esti/ugc_test.go [hunk 7]
```diff
@@ -214,16 +220,177 @@ func listUnderlyingStorage(t *testing.T, ctx context.Context, s3Client *s3.S3, b
...
 	// setup new s3 client for direct access to the underlying storage
 	mySession := session.Must(session.NewSession())
 	s3Client := s3.New(mySession, aws.NewConfig().WithRegion("us-east-1"))
 	return s3Client
 }
+
+func getLastUGCRunID(t *testing.T, ctx context.Context, s3Client *s3.S3, bucket, prefix string) string {
+	runIDPrefix := prefix + "_lakefs/retention/gc/uncommitted/"
+
+	listInput := &s3.ListObjectsInput{
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/ugc_test.go:247]: done

File: esti/ugc_test.go [hunk 8]
```diff
@@ -214,16 +220,177 @@ func listUnderlyingStorage(t *testing.T, ctx context.Context, s3Client *s3.S3, b
...
+		t.Fatalf("Failed to delete object %s", err)
+	}
+}
+
+func validateSafeUncommittedGC(t *testing.T, ctx context.Context, s3Client *s3.S3, repository string) {
+	objects, qp := listRepositoryUnderlyingStorage(t, ctx, repository)
+	runID := getLastUGCRunID(t, ctx, s3Client, qp.StorageNamespace, qp.Prefix)
+
+	reportPath := fmt.Sprintf("%s_lakefs/retention/gc/uncommitted/%s/summary.json", qp.Prefix, runID)
+	startListTime, err := getReportCutoffTime(s3Client, qp.StorageNamespace, reportPath)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/ugc_test.go:345]: done

File: esti/ugc_test.go [hunk 9]
```diff
@@ -214,16 +220,177 @@ func listUnderlyingStorage(t *testing.T, ctx context.Context, s3Client *s3.S3, b
...
+	ticker := time.NewTicker(time.Second * 1)
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case <-done:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				uploadAndDeleteSafeTestData(t, ctx, repo)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/ugc_test.go:293]: for the ugc to delete it

File: esti/ugc_test.go [hunk 10]
```diff
@@ -214,16 +220,177 @@ func listUnderlyingStorage(t *testing.T, ctx context.Context, s3Client *s3.S3, b
...
+		extraSubmitArgs: []string{"--conf", "spark.hadoop.lakefs.debug.gc.uncommitted_min_age_seconds=1"},
+		programArgs:     []string{uncommittedSafeGCRepoName, "us-east-1"},
+		logSource:       "ugc",
+	}
+	testutil.MustDo(t, "run uncommitted GC", runSparkSubmit(submitConfig))
+	done <- true
+
+	// post run
+	const expectedSizeBytes = 38
+	_, err = client.LinkPhysicalAddressWithResponse(ctx, "repo1", "main", &api.LinkPhysicalAddressParams{
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/ugc_test.go:312]: done

File: esti/ugc_test.go [hunk 11]
```diff
@@ -214,16 +220,177 @@ func listUnderlyingStorage(t *testing.T, ctx context.Context, s3Client *s3.S3, b
...
+		Path: "foo/bar",
+	}, api.LinkPhysicalAddressJSONRequestBody{
+		Checksum:  "afb0689fe58b82c5f762991453edbbec",
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/ugc_test.go:315]: example from the tests

**Review by itaiad200 — APPROVED** [review]

File: esti/ugc_test.go [hunk 12]
```diff
@@ -10,26 +11,31 @@ import (
 	"strconv"
 	"strings"
 	"testing"
+	"time"
+
+	"github.com/spf13/viper"
+	"github.com/treeverse/lakefs/pkg/config"
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/ugc_test.go:18]: done

**Pull request merged**